src/Disciplina/Educator/SizedMerkleTree.hs

```diff
@@ -0,0 +1,237 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveAnyClass #-}
+
+-- | Sized Merkle tree implementation.
+module Disciplina.Educator.SizedMerkleTree
+       ( SizedMerkleRoot(..)
+       , SizedMerkleTree (..)
+       , getSizedMerkleRoot
+       , mkSizedMerkleTree
+
+       , SizedMerkleNode (..)
+       , mkBranch
+       , mkLeaf
+       ) where
+
+import Universum hiding (foldMap, toList)
+
+import Disciplina.Crypto (Hash, HasHash, hash, unsafeHash)
+import Disciplina.Crypto.Hash.Class (HashFunc, AbstractHash (..))
+
+import Codec.Serialise (Serialise(..), serialise)
+import Data.Bits (Bits (..))
+import Data.ByteArray (ByteArrayAccess, convert)
+import Data.ByteString.Builder (Builder, byteString, lazyByteString)
+import Data.Foldable (Foldable (..))
+import Prelude (show)
+
+import qualified Data.ByteString.Builder.Extra as Builder
+import qualified Data.ByteString.Lazy as LBS
+
+-- | Data type for root of sized merkle tree.
+data SizedMerkleRoot a = SizedMerkleRoot
+    { smrHash :: Hash LBS.ByteString  -- ^ returns root 'Hash' of Merkle Tree
+    , smrSize :: Word8 -- ^ size of root node,
```

Comment:
This doesn't look like a "root". This is more like info of the tree.

Reply:
`SizedMerkleRoot` has nothing to do with the merkle tree root, so it should be renamed, I agree. Also, I would kindly ask you to arrange definitions in a readable inductive way -- this is the first datatype introduced and it's only used in the `SizedMerkleNode` which is defined after it's used in `SizedMerkleTree`.

Reply:
The term "root" is usually connected to node itself. How about "signature"?